### ISSUE TITLE
On Magento 2.3, form_key cookie expires immediately if cookie lifetime is set to 0

### ISSUE THREAD

**Issue by staradzinau** [magento/magento2#19718]
### Preconditions (*)

1. Magento 2.3.0
2. Steps for reproduce are exactly the same as mentioned in https://github.com/magento/magento2/issues/10527

### Additional information 
Commit https://github.com/magento/magento2/commit/99a3c012fd86cd1a4709413db48993f86e115549 wasn't correctly merged to the actual 2.3.0 brach. Please, compare computeExpirationTime() function:
https://github.com/magento/magento2/blob/2.2.7/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php
and 
https://github.com/magento/magento2/blob/2.3.0/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php

**Comment by magento-engcom-team** [comment]
Hi @avgust1111. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@avgust1111 do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by orlangur** [comment]
@avgust1111 such observations may be created directly as pull request ;) Thanks for your report.

**Comment by TomashKhamlai** [comment]
@magento-engcom-team give me 2.3.0 instance

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for your request. I'm working on Magento 2.3.0 instance for you

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai, here is your [Magento instance](https://i-19718-2-3-0.instances.magento-community.engineering).
Admin access: [https://i-19718-2-3-0.instances.magento-community.engineering/admin](https://i-19718-2-3-0.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by TomashKhamlai** [comment]
@avgust1111, thank you for report. I saw a label Fixed in 2.3.0 and also I was not able to reproduce issue by steps given.

**Comment by staradzinau** [comment]
@TomashKhamlai , please, note: I'm talking specifically about the commit 99a3c01 and its changes, which was merged into the 2.2.X branch, but it wasn't merged correctly into the 2.3.X branch. 
Because of this problem with the commit, we still have to hardcode a logic of the cookie expiration settings in order to make it work both on 2.2.X and 2.3.X.
The bug from the old issue (#10527) cannot be reproduced because of some other changes, but the initial problem (an impossibility to set a cookie which expires at the end of session time programmatically) still persists in the 2.3.X branch, despite it works properly in 2.2.7
Kindly review the difference in the `computeExpirationTime()` function
https://github.com/magento/magento2/blob/2.2.7/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php:
2.2.7:
```
if (isset($metadataArray[CookieMetadata::KEY_DURATION])
                && $metadataArray[CookieMetadata::KEY_DURATION] !== PhpCookieManager::EXPIRE_AT_END_OF_SESSION_TIME
            ) {
```
and
https://github.com/magento/magento2/blob/2.3.0/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php:
2.3.0:
`if (isset($metadataArray[CookieMetadata::KEY_DURATION])) {`

As you can see, it's necessary to add at least the second part of condition, e.g:
`&& $metadataArray[CookieMetadata::KEY_DURATION] !== PhpCookieManager::EXPIRE_AT_END_OF_SESSION_TIME`

This is the problem I'm reporting

**Comment by TomashKhamlai** [comment]
We will check it again.

**Comment by staradzinau** [comment]
@TomashKhamlai , thanks, I will be waiting for updates

**Comment by sivaschenko** [comment]
Hi @avgust1111 , thanks for your report. Can you please provide the actual steps to reproduce, actual and expected result for us to verify the issue.

**Comment by staradzinau** [comment]
Hello, @sivaschenko 
In fact, I didn't face with such an issue in the native Magento workflow. During extension developing, I've reproduced an impossibility to set a cookie which expires at the end of session time on 2.3.X programmatically in the correct way, without hardcoded logic.
Here are exact details. 
- Create an instance of `Magento\Framework\Stdlib\Cookie\PublicCookieMetadata` class
- Set all required parameters, and call `setDuration()` function with `$duration = Magento\Framework\Stdlib\Cookie\PhpCookieManager::EXPIRE_AT_END_OF_SESSION_TIME`
- Set some public cookie using `Magento\Framework\Stdlib\CookieManagerInterface` and its function `setPublicCookie()` with `PublicCookieMetadata` instance from previous steps

Actual result for 2.3.X: specific cookie is not set
Expected result: public cookie, which expires at the end of session time is set correctly

Please, investigate such an issue on 2.3.X.

**Comment by cscd98** [comment]
Can confirm this bug in 2.3.0..

**Comment by ghost** [comment]
@craigcarnell, did you reproduced the issue on clean Magento 2.3.0 or on Magento 2.3.0 upgraded from 2.2.x?

**Comment by cscd98** [comment]
Upgraded, I don't have time to install Magento for every single bug I find :)

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii, here is your [Magento instance](https://i-19718-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-19718-2-3-develop.instances.magento-community.engineering/admin](https://i-19718-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
@magento-engcom-team give me 2.3.0 instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for your request. I'm working on Magento 2.3.0 instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii, here is your [Magento instance](https://i-19718-2-3-0.instances.magento-community.engineering).
Admin access: [https://i-19718-2-3-0.instances.magento-community.engineering/admin](https://i-19718-2-3-0.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
@avgust1111, @craigcarnell could you please copy the whole function `function updateCookieFormKey` from `app/code/Magento/PageCache/Plugin/RegisterFormKeyFromCookie.php` or `vendor/magento/module-page-cache/Plugin/RegisterFormKeyFromCookie.php` and paste it to the comment?

**Comment by magento-engcom-team** [comment]
@engcom-backlog-andrii Thank you for verifying the issue.

Unfortunately, not enough information was provided to created internal ticket. Please consider adding the following:
- [ ] Update issue content to follow format required by [Issue Reporting Guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by ghost** [comment]
@avgust1111, please update the issue content to follow format required by [Issue Reporting Guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)

**Comment by ghost** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii, here is your [Magento instance](https://i-19718-2-2-develop.instances.magento-community.engineering).
Admin access: [https://i-19718-2-2-develop.instances.magento-community.engineering/admin](https://i-19718-2-2-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
@magento-engcom-team give me 2.2.7 instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for your request. I'm working on Magento 2.2.7 instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii, here is your [Magento instance](https://i-19718-2-2-7.instances.magento-community.engineering).
Admin access: [https://i-19718-2-2-7.instances.magento-community.engineering/admin](https://i-19718-2-2-7.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by cscd98** [comment]
@TomashKhamlai Why was this closed?? This is a major bug, I have to login to the admin several times..

**Comment by ghost** [comment]
@craigcarnell, I can reproduce it only on 2.1-develop. Please try to reproduce this problem in the provided instance.

**Comment by ghost** [comment]
@magento-engcom-team give me 2.3.0 instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-tomash. Thank you for your request. I'm working on Magento 2.3.0 instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-tomash, here is your [Magento instance](https://i-19718-2-3-0.instances.magento-community.engineering).
Admin access: [https://i-19718-2-3-0.instances.magento-community.engineering/admin](https://i-19718-2-3-0.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by cscd98** [comment]
@TomashKhamlai i've found my issue, might not be related to this, will create a ticket.
